### ISSUE TITLE
Make derivatives accessible via Style

### ISSUE THREAD

**Issue by keenancrane** [penrose/penrose#395]
For debugging purposes, it would be great to have derivatives of any variable available in Style.  This way, you can (say) visualize the direction the optimizer thinks it should move some object in order to decrease the energy / satisfy the constraints.  A mockup would be something like this:

```
forall Point p {
   // p's local
   p.x = ?
   p.y = ?

   // as usual define shape, etc., for p, in terms of p.x and p.y
   ...

   // DEBUG: draw an arrow showing where the optimizer wants to move p
   p.debugShape = Arrow {
      startX: p.x
      startY: p.y
      endX: p.x + derivative( p.x )
      endX: p.y + derivative( p.y )
   }
}
```

Here, **derivative** is a proposed Style keyword that substitutes the given variable for the derivative used by the optimizer.  More precisely, if the optimizer is minimizing an overall objective function F(x1, ..., xn), then `derivative( xi )` yields the current value of ∂F/∂xi.

Further thoughts:

- It might be useful to get the partial derivatives ∂F/∂x, _or_ the preconditioned descent direction for that variable.  E.g., we currently apply L-BFGS to get the actual descent direction, and there may be cases where the programmer wants to visualize this descent direction _or_ the raw gradient direction.
- Probably simplest to not allow arbitrary expressions inside a `derivative` directive.  Just a raw value substitution.  If one wants derivatives of a composite expression, they can build it up in Style and then pass it into `derivative`
- The idea of including derivatives in a graphics DSL is not without precedent—it's actually an important part of GLSL.  Here's the [man page](https://www.khronos.org/registry/OpenGL-Refpages/gl4/html/dFdx.xhtml).

**Comment by hypotext** [comment]
Great idea! I think it will really help Penrose be more of a "take-out" system (e.g. you can help yourself). I just implemented this on branch `opt-debug`, 98c61ed. Want to try it out and let me know how it works?

Example:

`runpenrose set-theory-domain/twosets-simple.sub set-theory-domain/venn-small.sty set-theory-domain/setTheory.dsl`

![twosets](https://user-images.githubusercontent.com/2762356/98148116-d444af00-1e9a-11eb-93ce-54596e7b4943.gif)

(derivative of text position shown in red, preconditioned derivative shown in blue)

Style: (note the two debug arrows)

```
forall Set x {
    x.shape = Circle {
        x : ?
        y : ?
        r : ?
        -- r : 100.0
        strokeWidth : 0
    }

    x.text = Text {
        string : x.label
    }

    LOCAL.c = 20.0

    x.debugShape = Arrow {
        startX: x.text.x
        startY: x.text.y
        endX: x.text.x + LOCAL.c * derivative(x.text.x)
        endY: x.text.y + LOCAL.c * derivative(x.text.y)
        color: rgba(1.0, 0.0, 0.0, 0.5)
        arrowheadSize: 0.5
        thickness: 3.0
        style: "dashed"
    }

    LOCAL.d = 20.0

    x.debugShape2 = Arrow {
        startX: x.text.x
        startY: x.text.y
        endX: x.text.x + LOCAL.d * derivativePreconditioned(x.text.x)
        endY: x.text.y + LOCAL.d * derivativePreconditioned(x.text.y)
        color: rgba(0.0, 0.0, 1.0, 0.5)
        arrowheadSize: 0.5
        thickness: 3.0
        style: "dashed"
    }

    ensure contains(x.shape, x.text) -- NEW
    encourage sameCenter(x.text, x.shape) -- OLD

    ensure minSize(x.shape)
    ensure maxSize(x.shape)
}
```

Notes:

* This only works well with a lower number of optimization steps per display steps, because the gradient info is only gotten every time the diagram is rendered. Currently it's 2 optimization steps per display step.

**Comment by hypotext** [comment]
Merged into master in f01cbafa1643df935c511ac1b3522a923832a3ce
